core/src/main/java/com/graphhopper/search/EdgeKVStorage.java

```diff
@@ -70,57 +83,61 @@ public EdgeKVStorage create(long initBytes) {
         return this;
     }
 
     public boolean loadExisting() {
         if (vals.loadExisting()) {
             if (!keys.loadExisting())
                 throw new IllegalStateException("Loaded values but cannot load keys");
-            int stringIndexKeysVersion = keys.getHeader(0);
-            int stringIndexValsVersion = vals.getHeader(0);
-            GHUtility.checkDAVersion(keys.getName(), Constants.VERSION_STRING_IDX, stringIndexKeysVersion);
-            GHUtility.checkDAVersion(vals.getName(), Constants.VERSION_STRING_IDX, stringIndexValsVersion);
-            bytePointer = BitUtil.LITTLE.combineIntsToLong(vals.getHeader(4), vals.getHeader(8));
+            bytePointer = bitUtil.combineIntsToLong(vals.getHeader(0), vals.getHeader(4));
 
             // load keys into memory
             int count = keys.getShort(0);
             long keyBytePointer = 2;
             for (int i = 0; i < count; i++) {
                 int keyLength = keys.getShort(keyBytePointer);
                 keyBytePointer += 2;
-
                 byte[] keyBytes = new byte[keyLength];
                 keys.getBytes(keyBytePointer, keyBytes, keyLength);
                 String valueStr = new String(keyBytes, Helper.UTF_CS);
-                keysInMem.put(valueStr, keysInMem.size());
-                keyList.add(valueStr);
                 keyBytePointer += keyLength;
+
+                keyToIndex.put(valueStr, keyToIndex.size());
+                indexToKey.add(valueStr);
+
+                int shortClassNameLength = 1;
+                byte[] classBytes = new byte[shortClassNameLength];
+                keys.getBytes(keyBytePointer, classBytes, shortClassNameLength);
+                keyBytePointer += shortClassNameLength;
+                indexToClass.add(shortNameToClass(new String(classBytes, Helper.UTF_CS)));
             }
             return true;
         }
 
         return false;
     }
 
-    Set<String> getKeys() {
-        return keysInMem.keySet();
+    Collection<String> getKeys() {
+        return indexToKey;
     }
 
     /**
-     * This method writes the specified key-value pairs into the storage.
+     * This method writes the specified key-value pairs into the storage. Please note that null keys or null values
+     * are rejected.
      *
      * @return entryPointer to later fetch the entryMap via get
      */
-    public long add(Map<String, String> entryMap) {
+    public long add(final Map<String, Object> entryMap) {
         if (entryMap.isEmpty())
             return EMPTY_POINTER;
         else if (entryMap.size() > 200)
             throw new IllegalArgumentException("Cannot store more than 200 entries per entry");
 
-        // This is a very important compressing mechanism due to the nature of OSM. Make this faster via precalculated hastcodes?
+        // This is a very important "compression" mechanism due to the nature of OSM.
         if (entryMap.equals(lastEntryMap))
             return lastEntryPointer;
```

Comment:
It looks like this compression relied on `String#equals` or the fact that Strings are immutable. But does it still work when the values are arbitrary `Object`?  It seems like the objects need to implement some kind of object equality so we recognize different objects with the same contents as 'equal' here. Of course this is fulfilled for String and also Integer, Long, Double etc., but maybe it is worth a comment/javadoc or something? Or are the Objects always expected to be a `String` or a `Number`? Then we could even enforce this here.

Reply:
Oh, right. But why do we have to parse anything then? Can we not just check `object instanceof String || object instanceof Number || object instanceof byte[]`?

Reply:
Btw for byte[] the equals method probably does not behave as we'd like it to behave. We would have to use `Arrays.equals`:

```java
    @Test
    public void byteArrayEquals() {
        byte[] a = new byte[] { 0, 1, 2};
        byte[] b = new byte[] { 0, 1, 2};
        assertFalse(a.equals(b));
        assertTrue(Arrays.equals(a, b));
    }
```

Reply:
If a method accepts a `Map<String, Object>` and `Object` can only be a `Number`, `String` or a `byte[]` I think it can improve readability when the first thing this method does is checking if that is actually the case. Especially when we run lots of code before this differentiation is done (like everything before the first time we call getBytesForValue here(?)). So yes there is no real need to do it here, but since we already distinguish the `byte[]` case it would not hurt to check for String/Number as well. I would not insist on this though :)